src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Engine/AbstractFormatEngine.cs

```diff
@@ -162,13 +162,17 @@ protected virtual NodeOperations CreateNodeOperations(CancellationToken cancella
             return nodeOperations;
         }
 
-        private static void AddOperations<T>(List<T> operations, List<T> scratch, SyntaxNode node, Action<List<T>, SyntaxNode> addOperations)
+        private static void AddOperations<T>(SegmentedList<T> operations, List<T> scratch, SyntaxNode node, Action<List<T>, SyntaxNode> addOperations)
         {
             Debug.Assert(scratch.Count == 0);
 
             addOperations(scratch, node);
+            foreach (var operation in scratch)
+            {
+                if (operation is not null)
```

Comment:
might be nicer if nulls can't get in in the first place.  basically, call out to the helpers to give back the operations, and if they return null, they never go into the lists.

Reply:
Those helpers are in the formattingrules class. Those methods appear to be virtual, so I don't really want to change them or assume that they aren't putting null in these array.